### PR TITLE
Union

### PR DESCRIPTION
**Pull request by WirajanDASILVA**
* **Please check if the PR fulfills these requirements**
- [ ] The commit messages follows our guidelines (explicit, in english)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [ ] CHANGELOG.md has been updated

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
- [X] Bug fix:
- [X] New features:
- [ ] Documentation
- [ ] Optimization

* **Other information**:

### CODE REVIEW

File: volmdlr/core.py [hunk 1]
```diff
@@ -666,9 +666,9 @@ def bbox_intersection(self, bbox2):
                 and self.zmin < bbox2.zmax and self.zmax > bbox2.zmin)
 
     def is_inside_bbox(self, bbox2):
-        return (self.xmin > bbox2.xmin and self.xmax < bbox2.xmax \
-                and self.ymin > bbox2.ymin and self.ymax < bbox2.ymax \
-                and self.zmin > bbox2.zmin and self.zmax < bbox2.zmax)
+        return (round(self.xmin, 6) >= round(bbox2.xmin, 6) and round(self.xmax, 6) <= round(bbox2.xmax, 6) \
+                and round(self.ymin, 6) >= round(bbox2.ymin, 6) and round(self.ymax, 6) <= round(bbox2.ymax, 6) \
+                and round(self.zmin, 6) >= round(bbox2.zmin, 6) and round(self.zmax, 6) <= round(bbox2.zmax, 6))
```
Comment (masfaraud) [inline volmdlr/core.py:671]: better use math.isclose(val1, val2, abs_tol=1e-6)

File: volmdlr/edges.py [hunk 2]
```diff
@@ -674,6 +676,11 @@ def middle_point(self):
 
     def point_at_abscissa(self, curvilinear_abscissa):
         return self.start + self.unit_direction_vector() * curvilinear_abscissa
+    
+    def point_belongs(self, point):
+        if self.start.point_distance(point) + self.end.point_distance(point) == self.length():
+            return True
+        return False
```
Comment (masfaraud) [inline volmdlr/edges.py:683]: I don't think this is correct

File: volmdlr/edges.py [hunk 3]
```diff
@@ -1952,6 +1962,10 @@ def length(self):
     def point_at_abscissa(self, curvilinear_abscissa):
         return self.start + curvilinear_abscissa * (
                 self.end - self.start) / self.length()
+    def point_belongs(self, point):
+        if self.start.point_distance(point) + self.end.point_distance(point) == self.length():
+            return True
+        return False
```
Comment (masfaraud) [inline volmdlr/edges.py:1968]: This is not right I guess. We could develop a __contains__method

File: volmdlr/edges.py [hunk 4]
```diff
@@ -2051,6 +2065,70 @@ def intersection(self, segment2):
...
+                vector1.normalize()
+                vector2.normalize()
+                if vector1 == self.direction_vector() and vector2 == linesegment.direction_vector() and \
+                    self.start.point_distance(intersection)+ self.end.point_distance(intersection) == self.length() and\
+                       linesegment.start.point_distance(intersection) + linesegment.end.point_distance(intersection) == linesegment.length() :
+                    return intersection
+                else: 
+                    return None
+        return None
+
```
Comment (masfaraud) [inline volmdlr/edges.py:2130]: Maybe use line_intersection and see if it is in segment

File: scripts/ci_scripts.py [hunk 5]
```diff
@@ -1,10 +1,13 @@
...
-           'wires/roundedlines.py','wires/polygon2D.py',
-           'wires/triangle2D.py',
-           'primitives/extrusion.py', 'demo2D.py', 'showcases/casing.py', 'primitives/sweep.py',
+scripts = [
+    # 'edges/arcs2D.py', 'edges/arcs3D.py', 'showcases/simple_shapes.py',
+    #        'wires/roundedlines.py','wires/polygon2D.py',
+    #        'wires/triangle2D.py',
+    #        'primitives/extrusion.py', 'demo2D.py', 
+    #        'showcases/casing.py', 
+           # 'primitives/sweep.py',
```
Comment (masfaraud) [inline scripts/ci_scripts.py:10]: Put back scripts

File: scripts/cloud/cloud_debug2.py [hunk 6]
```diff
@@ -481,177 +481,179 @@
...
-#             for face in faces :
-#                 vme.LineSegment3D(face.point1, face.point2).plot(ax=ax, color='g')
-#                 vme.LineSegment3D(face.point2, face.point3).plot(ax=ax,color='g')
-#                 vme.LineSegment3D(face.point3, face.point1).plot(ax=ax, color='g')
-#         #     count += 1
-#         #     break
-#         # if count > 0:
-#         #     break
-# #     vm
-
```
Comment (masfaraud) [inline scripts/cloud/cloud_debug2.py:511]: Clean commented parts

File: scripts/cloud/cloud_debug2.py [hunk 7]
```diff
@@ -481,177 +481,179 @@
...
+
+# points_poly1 = [vm.Point3D(-0.02785286331176758, 0.14365164184570312, 0.16285931396484374),
+# vm.Point3D(-0.039570194244384765, 0.14255133056640626, 0.16285931396484374),
+# vm.Point3D(-0.04901511764526367, 0.14049978637695312, 0.16285931396484374),
+# vm.Point3D(-0.05265781402587891, 0.138816650390625, 0.16285931396484374),
+# vm.Point3D(-0.05629143524169922, 0.1368720245361328, 0.16285931396484374),
+# vm.Point3D(-0.06060961151123047, 0.1323123321533203, 0.16285931396484374),
+# vm.Point3D(-0.06237499237060547, 0.12723016357421876, 0.16285931396484374),
+# vm.Point3D(-0.06580715942382813, 0.11276873016357422, 0.16285931396484374),
+# vm.Point3D(-0.06730508422851562, 0.09772169494628906, 0.16285931396484374),
```
Comment (masfaraud) [inline scripts/cloud/cloud_debug2.py:523]: Also clean

File: scripts/faces/triangulation_planeface.py [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+    # for face in file_faces.triangles:
+    #     points = face.triangulation().points
+    #     for three_pos in face.triangulation().triangles :
+    #         triangles.append(vmf.Triangle3D(points[three_pos[0]], 
+    #                                         points[three_pos[1]],
+    #                                         points[three_pos[2]]))
+  
+for triangles in list_tri:
+    print()          
+    for tri in triangles :
```
Comment (masfaraud) [inline scripts/faces/triangulation_planeface.py:27]: What is the point of this script? Show ZeroDivision Error?

File: scripts/faces/union.py [hunk 9]
```diff
@@ -0,0 +1,285 @@
...
+# # poly3_vol2.plot(ax=ax, color = 'g')
+# poly4_vol2 = poly3_vol2.rotation(vm.O3D, vm.Y3D, math.pi/4).translation(0.05*vm.Y3D)
+# # poly4_vol2.plot(ax=ax, color = 'r')
+# poly5_vol2 = poly4_vol2.rotation(vm.O3D, vm.Y3D, math.pi/10).translation(0.2*vm.Y3D)
+# # poly5_vol2.plot(ax=ax, color = 'y')
+
+# point_triangles_2 = poly1_vol2.sewing(poly2_vol2, vm.X3D, vm.Z3D) + poly2_vol2.sewing(poly3_vol2, vm.X3D, vm.Z3D) +\
+#                     poly3_vol2.sewing(poly4_vol2, vm.X3D, vm.Z3D) + poly4_vol2.sewing(poly5_vol2, vm.X3D, vm.Z3D)
+
+# faces_2 = [vmf.Triangle3D(trio[0], trio[1], trio[2]) for trio in point_triangles_2]
```
Comment (masfaraud) [inline scripts/faces/union.py:134]: Remove commented parts

File: volmdlr/cloud.py [hunk 10]
```diff
@@ -107,10 +108,11 @@ def to_shell(self, resolution = 10, normal = None):
                 faces.append(vmf.PlaneFace3D(plane3d, surf2d))
             if n != resolution-1:
                 poly2 = polygon3d[n+1]
+                poly2.plot(ax=ax, color='r')
                 # poly2 = poly2.simplify(0.01, 0.05)
                 
                 # coords = poly1.sewing_with(poly2, vec1, vec2, normal, resolution = max_poly_resolution)
-                coords = poly1.sewing(poly2)
+                coords = poly1.sewing(poly2, vec1, vec2)
                 for trio in coords :
```
Comment (masfaraud) [inline volmdlr/cloud.py:116]: Remove plots for production use?

File: volmdlr/core.py [hunk 11]
```diff
@@ -666,9 +666,12 @@ def bbox_intersection(self, bbox2):
...
     def is_inside_bbox(self, bbox2):
-        return (self.xmin > bbox2.xmin and self.xmax < bbox2.xmax \
-                and self.ymin > bbox2.ymin and self.ymax < bbox2.ymax \
-                and self.zmin > bbox2.zmin and self.zmax < bbox2.zmax)
+        return ((self.xmin >= bbox2.xmin or math.isclose(self.xmin, bbox2.xmin, abs_tol=10e-6))\
+                and (self.xmax <= bbox2.xmax or math.isclose(self.xmax, bbox2.xmax, abs_tol= 10e-6))\
+                and (self.ymin >=bbox2.ymin or math.isclose(self.ymin, bbox2.ymin, abs_tol= 10e-6))\
+                and (self.ymax <= bbox2.ymax or math.isclose(self.ymax, bbox2.ymax, abs_tol=10e-6)) \
+                and (self.zmin >= bbox2.zmin  or math.isclose(self.zmin, bbox2.zmin, abs_tol=10e-6))\
+                and (self.zmax <= bbox2.zmax or math.isclose(self.zmax, bbox2.zmax, abs_tol=10e-6)))
```
Comment (masfaraud) [inline volmdlr/core.py:674]: for perf better use a < b + 1e-6
And 10e-6 = 1e-05

File: volmdlr/faces.py [hunk 12]
```diff
@@ -2344,12 +2347,25 @@ def face_intersections(self, face2):
...
+        if intersections:
+            # print('len(intersections) :',  len(intersections))
+            try:
+                primitive = volmdlr.edges.LineSegment3D(intersections[0], intersections[1])
+                intersections = volmdlr.wires.Wire3D([primitive])
+                return intersections
+            except IndexError:# TODO: for debugging, remember to delete after 
+                # ax = self.plot(color = 'r')
+                # face2.plot(ax=ax)
+                return None
```
Comment (masfaraud) [inline volmdlr/faces.py:2369]: Better return a list of intersection rather than None or Value

File: volmdlr/faces.py [hunk 13]
```diff
@@ -4027,7 +4047,8 @@ def intersection_external_aabb_volume(self, shell2: 'OpenShell3D',
         for face1 in self.faces:
             for face2 in shell2.faces:
                 intersection_points = face1.face_intersections(face2)
-                if intersection_points is not None:
+                if intersection_points:
+                    intersection_points = [intersection_points.primitives[0].start, intersection_points.primitives[0].end]
```
Comment (masfaraud) [inline volmdlr/faces.py:4051]: Having a list of intersections would allow not to test for None

File: volmdlr/faces.py [hunk 14]
```diff
@@ -4144,16 +4165,35 @@ def copy(self):
...
+    #     return cls(shell1_p)
+    #     # shell2_p = shell2.shell_substract2(shell1, True)
+    #     # for f in shell1_p.faces:
+    #     #     f.alpha = 0.2
+    #     # for f in shell2_p.faces:
+    #     #     f.alpha = 0.5
+    #     #     f.color = (1, 0, 0)
+    #     # return cls(shell1_p.faces + shell2_p.faces)
+    #     # return cls(shell2_p.faces)
+
```
Comment (masfaraud) [inline volmdlr/faces.py:4198]: Clean

File: volmdlr/wires.py [hunk 15]
```diff
@@ -67,10 +67,35 @@ def point_at_abscissa(self, curvilinear_abscissa: float):
         raise ValueError(
             'abscissa over length: {}>{}'.format(curvilinear_abscissa, length))
 
-    def extract_primitives(self, point1, primitive1, point2, primitive2):
+    def extract_primitives(self, point1, primitive1, point2, primitive2, inter_points_contour):
```
Comment (masfaraud) [inline volmdlr/wires.py:70]: Doc for inter_points_contour and give a default value, typing also

File: volmdlr/wires.py [hunk 16]
```diff
@@ -2805,125 +3118,104 @@ def sewing_with(self, other_poly3d, x, y, normal, resolution = 20):
     def simplify(self, min_distance:float = 0.01, max_distance:float=0.05):
         return ClosedPolygon3D(self.simplify_polygon(min_distance = min_distance, max_distance = max_distance).points)
 
-    def sewing(self, polygon2):
+    def sewing(self, polygon2, x, y):
```
Comment (masfaraud) [inline volmdlr/wires.py:3121]: What are x and y? -> doc

File: volmdlr/wires.py [hunk 17]
```diff
@@ -2805,125 +3118,104 @@ def sewing_with(self, other_poly3d, x, y, normal, resolution = 20):
...
-        
-    #     lines = [volmdlr.edges.LineSegment2D(volmdlr.O2D, max_scale*(volmdlr.X2D*math.sin(n*2*math.pi/4) +
-    #                                          volmdlr.Y2D*math.cos(n*2*math.pi/4))
-    #                                          ) for n in range (4)]
-        
-    #     # center1, center2 = volmdlr.Point3D(center1.x, center1.y, 0), volmdlr.Point3D(center2.x, center2.y, 0)
-    #     # new_polygon1, new_polygon2 =self.translation(-center1), polygon2.translation(-center2)
-    #     # center1, center2 = new_polygon1.average_center_point(), new_polygon2.average_center_point()
-    #     # print(center1, center2)
-
```
Comment (masfaraud) [inline volmdlr/wires.py:2904]: Clean if unused

**Review by masfaraud — CHANGES_REQUESTED** [review]

File: scripts/faces/triangulation_planeface.py (hunk 8 above)
Comment (WirajanDASILVA) [reply scripts/faces/triangulation_planeface.py:27]: yes, that is correct. Its is a test case to show this error

File: volmdlr/core.py (hunk 11 above)
Comment (WirajanDASILVA) [reply volmdlr/core.py:674]: then we use math.isclose no more?

File: scripts/faces/triangulation_planeface.py (hunk 8 above)
Comment (masfaraud) [reply scripts/faces/triangulation_planeface.py:27]: as it is not a bug but a feature we can remoce this script and the face0.stl and face1.stl files

Comment (WirajanDASILVA) [reply scripts/faces/triangulation_planeface.py:27]: alright

**Review by masfaraud — APPROVED** [review]

**Review by masfaraud — APPROVED** [review]

**Pull request merged**